rust/connlib/libs/common/src/session.rs

```diff
@@ -0,0 +1,241 @@
+// so revisit this.
+/// A session is the entry-point for connlib, mantains the runtime and the tunnel.
+///
+/// A session is created using [Session::connect], then to stop a session we use [Session::disconnect].
+pub struct Session<T, U, V, R, M> {
+    runtime: Option<Runtime>,
+    _phantom: PhantomData<(T, U, V, R, M)>,
+}
+
+/// Resource list that will be displayed to the users.
+pub struct ResourceList {
+    pub resources: Vec<String>,
+}
+
+/// Tunnel addresses to be surfaced to the client apps.
+pub struct TunnelAddresses {
+    /// IPv4 Address.
+    pub address4: Ipv4Addr,
+    /// IPv6 Address.
+    pub address6: Ipv6Addr,
+}
+
+// Evaluate doing this not static
+/// Traits that will be used by connlib to callback the client upper layers.
+pub trait Callbacks {
+    /// Called when there's a change in the resource list.
+    fn on_update_resources(resource_list: ResourceList);
+    /// Called when the tunnel address is set.
+    fn on_set_tunnel_adresses(tunnel_addresses: TunnelAddresses);
+    /// Called when there's an error.
+    ///
+    /// # Parameters
+    /// - `error`: The actual error that happened.
+    /// - `error_type`: Wether the error should terminate the session or not.
+    fn on_error(error: &Error, error_type: ErrorType);
+}
+
+macro_rules! fatal_error {
+    ($result:expr, $c:ty) => {
+        match $result {
+            Ok(res) => res,
+            Err(e) => {
+                <$c>::on_error(&e, ErrorType::Fatal);
+                return;
+            }
+        }
+    };
+}
+
+impl<T, U, V, R, M> Session<T, U, V, R, M>
+where
+    T: ControlSession<M, V>,
+    U: for<'de> serde::Deserialize<'de> + std::fmt::Debug + Send + 'static,
+    R: for<'de> serde::Deserialize<'de> + std::fmt::Debug + Send + 'static,
+    V: serde::Serialize + Send + 'static,
+    M: From<U> + From<R> + Send + 'static + std::fmt::Debug,
+{
+    /// Block on waiting for ctrl+c to terminate the runtime.
+    /// (Used for the gateways).
+    pub fn wait_for_ctrl_c(&mut self) -> Result<()> {
+        self.runtime
+            .as_ref()
+            .ok_or(Error::NoRuntime)?
+            .block_on(async {
+                tokio::signal::ctrl_c().await?;
+                Ok(())
+            })
+    }
+
+    /// Starts a session in the background.
+    ///
+    /// This will:
+    /// 1. Create and start a tokio runtime
+    /// 2. Connect to the control plane to the portal
+    /// 3. Start the tunnel in the background and forward control plane messages to it.
+    ///
+    /// The generic parameter `C` should implement all the handlers and that's how errors will be surfaced.
+    ///
+    /// On a fatal error you should call `[Session::disconnect]` and start a new one.
+    // TODO: token should be something like SecretString but we need to think about FFI compatibiltiy
+    pub fn connect<C: Callbacks>(portal_url: impl TryInto<Url>, token: String) -> Result<Self> {
```

Comment:
So if I understand correctly, the reason we are using a callback-based design here is because of how connlib is integrated into the clients, right? Like, Android wants to be able to set callbacks on when certain events happen.

In my experience, callbacks are nice for application code but they don't compose very well, meaning for anything further down the stack, it often works well to implement stuff as state machines and only provide a callback-based interface at the top-level.